### PR TITLE
Add custom color campcollaboration

### PR DESCRIPTION
**Pull request by manuelmeister**
Followup PR after #5174

### CODE REVIEW

File: api/src/Entity/CampCollaboration.php [hunk 1]
```diff
@@ -202,6 +202,25 @@ class CampCollaboration extends BaseEntity implements BelongsToCampInterface {
...
+    public ?string $color = null;
+
+    /**
+     * The abbreviation in th avatar.
+     */
+    #[InputFilter\Trim]
+    #[ApiProperty(example: 'AB')]
+    #[Groups(['read', 'write'])]
+    #[ORM\Column(type: 'string', length: 2, nullable: true)]
+    public ?string $abbr = null;
```
Comment (usu) [inline api/src/Entity/CampCollaboration.php]: I'm not a big fan of abbr.s 😄 
Would rather use the full name `abbreviation`

Do we want to limit to type of characters? Or do we keep it open by purpose (e.g. allowing emojis)?

File: api/src/Entity/CampCollaboration.php [hunk 2]
```diff
@@ -202,6 +202,25 @@ class CampCollaboration extends BaseEntity implements BelongsToCampInterface {
...
+     */
+    #[InputFilter\Trim]
+    #[Assert\Regex(pattern: '/^#[0-9a-zA-Z]{6}$/')]
+    #[ApiProperty(example: '#4DBB52')]
+    #[Groups(['read', 'write'])]
+    #[ORM\Column(type: 'string', length: 8, nullable: true)]
+    public ?string $color = null;
+
+    /**
+     * The abbreviation in th avatar.
```
Comment (usu) [inline api/src/Entity/CampCollaboration.php]:
```suggestion
     * The abbreviation in the avatar.
```

File: frontend/src/components/collaborator/CollaboratorCreate.vue [hunk 3]
```diff
@@ -51,15 +51,17 @@ export default {
   },
   data() {
     return {
-      entityProperties: ['camp', 'inviteEmail', 'role'],
+      entityProperties: ['abbr', 'camp', 'color', 'inviteEmail', 'role'],
```
Comment (usu) [inline frontend/src/components/collaborator/CollaboratorCreate.vue]: I would rather not include abbreviation and color in the create dialog.
Limit inputs fields during creation, I can still go back and edit/customize if I want to.

**Review by usu — COMMENTED** [review]
Cool 🚀

File: api/src/Entity/CampCollaboration.php (hunk 1 above)
Comment (manuelmeister) [reply api/src/Entity/CampCollaboration.php]: Exactly. I thought that would be a fun case, and I'm not sure how I can limit it, without preventing certain emojis.

**Comment by manuelmeister** [comment]
## Core Meeting Decision

Implement as request by @usu

File: api/src/Entity/CampCollaboration.php [hunk 4]
```diff
@@ -213,17 +213,14 @@ class CampCollaboration extends BaseEntity implements BelongsToCampInterface {
...
     #[InputFilter\Trim]
+    #[Assert\Length(max: 2, countUnit: Assert\Length::COUNT_GRAPHEMES)]
     #[ApiProperty(example: 'AB')]
     #[Groups(['read', 'write'])]
-    #[ORM\Column(type: 'string', length: 2, nullable: true)]
-    public ?string $abbr = null;
-
-    #[ApiProperty(readable: false, writable: false)]
     #[ORM\Column(type: 'text', nullable: true)]
-    public ?string $collaborationAcceptedBy = null;
```
Comment (usu) [inline api/src/Entity/CampCollaboration.php:207]: Why was this removed? Is was not used?

**Review by usu — APPROVED** [review]

File: api/src/Entity/CampCollaboration.php (hunk 4 above)
Comment (manuelmeister) [reply api/src/Entity/CampCollaboration.php:207]: It was fully dead code. Unused, and not even tested.

File: common/helpers/campCollaborationInitials.js [hunk 5]
```diff
@@ -5,5 +5,7 @@ import initials from './initials.js'
  * Returns two characters to display for a camp collaboration based on its user
  */
 export default function (campCollaboration) {
-  return initials(campCollaborationDisplayName(campCollaboration, null, false))
+  return campCollaboration?.abbreviation
```
Comment (carlobeltrame) [inline common/helpers/campCollaborationInitials.js:8]: Could you add a test for this new behaviour in our existing `__tests__/campCollaborationInitials.spec.js`?

File: common/helpers/colors.js [hunk 6]
```diff
@@ -58,6 +58,10 @@ function campCollaborationColor(campCollaboration) {
     (typeof campCollaboration.user === 'function' &&
       campCollaboration.user()._meta?.loading)
 
+  if (campCollaboration?.color) {
```
Comment (carlobeltrame) [inline common/helpers/colors.js:61]: Test for this as well?

File: frontend/src/plugins/veeValidate/__tests__/oneEmojiOrTwoCharacters.spec.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+const mockI18n = {
+  $tc: (key) => key,
+}
+
+describe('oneEmojiOrTwoCharacters validation', () => {
+  it.each([
+    ['1', true],
+    ['12', true],
+    ['123', false],
+    ['🧑🏼‍🔧', true],
```
Comment (carlobeltrame) [inline frontend/src/plugins/veeValidate/__tests__/oneEmojiOrTwoCharacters.spec.js:12]: Does one of these tests check for combining emojis? Such as skin color combiners etc.. I'm not an expert on this, so I don't know whether one of your tests already accounts for that.

File: frontend/src/plugins/veeValidate/oneEmojiOrTwoCharacters.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import { size } from 'lodash'
+
+export default (i18n) => ({
+  validate: (value) => {
+    return /\p{Extended_Pictographic}/u.test(value) ? size(value) <= 1 : value.length <= 2
```
Comment (carlobeltrame) [inline frontend/src/plugins/veeValidate/oneEmojiOrTwoCharacters.js:5]: Wow, I learned something today.
In https://stackoverflow.com/a/72727900 it says that extended pictographic does not match a lone snowflake emoji, is that really the case? Should we use `(\p{Emoji}\uFE0F|\p{Emoji_Presentation})` as that post suggests? Or is that covered in your additional logic with the length?

File: api/src/Entity/CampCollaboration.php [hunk 9]
```diff
@@ -202,9 +202,25 @@ class CampCollaboration extends BaseEntity implements BelongsToCampInterface {
...
+    #[ApiProperty(example: '#4DBB52')]
+    #[Groups(['read', 'write'])]
+    #[ORM\Column(type: 'string', length: 8, nullable: true)]
+    public ?string $color = null;
+
+    /**
+     * The abbreviation in the avatar.
+     */
+    #[InputFilter\Trim]
+    #[Assert\Length(max: 2, countUnit: Assert\Length::COUNT_GRAPHEMES)]
```
Comment (carlobeltrame) [inline api/src/Entity/CampCollaboration.php:219]: This will allow two emoji, correct? Not that it's a huge issue, but wouldn't COUNT_CODEPOINTS be more correct for these specific rules which we are enforcing? (1 emoji or 2 characters) https://symfony.com/doc/current/reference/constraints/Length.html#countunit

File: frontend/src/plugins/veeValidate/__tests__/oneEmojiOrTwoCharacters.spec.js (hunk 7 above)
Comment (manuelmeister) [reply frontend/src/plugins/veeValidate/__tests__/oneEmojiOrTwoCharacters.spec.js:12]: This is already one that is made out of 3 combined emojis.

> The Man Mechanic: Light Skin Tone emoji is a [ZWJ sequence](https://emojipedia.org/emoji-zwj-sequence) combining [👨 Man](https://emojipedia.org/man), [🏻 Light Skin Tone](https://emojipedia.org/light-skin-tone), [Zero Width Joiner](https://emojipedia.org/zero-width-joiner) and [🔧 Wrench](https://emojipedia.org/wrench).
– Emojipedia

File: frontend/src/plugins/veeValidate/oneEmojiOrTwoCharacters.js (hunk 8 above)
Comment (manuelmeister) [reply frontend/src/plugins/veeValidate/oneEmojiOrTwoCharacters.js:5]: Yes it is covered.

File: api/src/Entity/CampCollaboration.php (hunk 9 above)
Comment (manuelmeister) [reply api/src/Entity/CampCollaboration.php:219]: Nope, an emoji could be made out of more than two codepoints, that's the reason we use graphemes. But because we want to allow 2 letters as well, we need to allow 2.

File: frontend/src/plugins/veeValidate/__tests__/oneEmojiOrTwoCharacters.spec.js (hunk 7 above)
Comment (manuelmeister) [reply frontend/src/plugins/veeValidate/__tests__/oneEmojiOrTwoCharacters.spec.js:12]: That's also the reason, that I've included some basic smileys below that should be each counted as one.

**Review by carlobeltrame — APPROVED** [review]
Looks great, thanks! As a future extension, I would like to be able to always change the color and abbreviation of my own camp collaboration, even if I'm only a guest. Would you agree?

**Comment by manuelmeister** [comment]
@carlobeltrame It somehow doesn't work with clientprint 😢  I didn't test it with a composed emoji until now. With "normal" emojis it works.

**Comment by manuelmeister** [comment]
@carlobeltrame I think we may rather add a color and abbreviation on the user, instead of user editable camp collaborators.

**Comment by carlobeltrame** [comment]
> @carlobeltrame It somehow doesn't work with clientprint 😢 I didn't test it with a composed emoji until now. With "normal" emojis it works.

I see, but it looks like composed emoji are a general problem in client print. No reason to delay merging this feature in my opinion.

**Comment by manuelmeister** [comment]
@carlobeltrame indeed, emojis are very slow with clientprint 😢. It doesn't break, but initially it needs to load all emojis, and this is probably taking a while.

**Comment by carlobeltrame** [comment]
> @carlobeltrame I think we may rather add a color and abbreviation on the user, instead of user editable camp collaborators.

That would work as well (though not as flexible). But if that's the ultimate goal, why didn't you implement that in the first place?

**Comment by carlobeltrame** [comment]
> @carlobeltrame indeed, emojis are very slow with clientprint 😢. It doesn't break, but initially it needs to load all emojis, and this is probably taking a while.

If the loading time is the biggest problem, we could preload the emojis as soon as you navigate to a page with a print button. But also, currently, the combined emoji are rendered split up into their components. I think that is actually an intended fallback in the unicode standard, so it's okay but just not optimal.

**Comment by manuelmeister** [comment]
> That would work as well (though not as flexible). But if that's the ultimate goal, why didn't you implement that in the first place?

Because it serves another purpose. I wanted a way to distinguish different users in a camp.

The user color/abbreviation would just be an override between generated and campcollaboration color/abbreviation, and be a better fallback.
